### PR TITLE
Fix overlap so that ``set_index`` doesn't drop rows

### LINKED ISSUE #8999: Results of dask.multi.merge_asof depends on npartitions
**What happened**: Results of `merge_asof` differ depending on npartitions of the dataframe. In particular, it seems the result misses rows when the number of partitions is equal to or greater than the number of rows in the dataframe. (I am aware that setting the number of partitions higher than row count does not make sense, yet I did not expect it to lead to this behavior). 

**What you expected to happen**: Identical results of `merge_asof` independent from npartitions of the dataframe. Or, if this issue is restricted to the case where npartitions>number of rows, an exception or warning telling the user that it may lead to undefined behavior.

**Minimal Complete Verifiable Example**:

```python
import pandas as pd
import dask.dataframe as dd

df = pd.DataFrame(dict(ts=[pd.to_datetime('1-1-2020')]*3, foo=[1,2,3]))
    
for npartitions in range(1,5):
    ddf = dd.from_pandas(df, npartitions=npartitions)
    result = dd.multi.merge_asof(left=ddf, right=ddf, on='ts').compute()
    print(result)

# Output
#          ts  foo_x  foo_y
# 0 2020-01-01      1      3
# 1 2020-01-01      2      3
# 2 2020-01-01      3      3
#           ts  foo_x  foo_y
# 0 2020-01-01      1      3
# 1 2020-01-01      2      3
# 2 2020-01-01      3      3
#           ts  foo_x  foo_y
# 0 2020-01-01      2      3
# 1 2020-01-01      3      3
#           ts  foo_x  foo_y
# 0 2020-01-01      2      3
# 1 2020-01-01      3      3
```
**What else do we need to know**: As far as I could verify, the issue only appears when the timestamps in the `on` column are identical.

**Environment**:

- Dask version: 2022.04.1
- Python version: 3.9.1
- Operating System: Debian GNU/Linux 11 (bullseye)
- Install method (conda, pip, source): pip

### LINKED ISSUE #9339: `ddf.set_index` with `sorted=True` drops rows
Hi. I see the same issue, but I don't have empty partitions. Can someone explain why I lost records after `set_index`?
```
df = pd.DataFrame({
    'ts': [1, 1, 2, 2, 3, 3, 3, 3],
    'value': 'abc'
})
ddf = dd.from_pandas(df, npartitions=3)
ddf = ddf.set_index('ts', sorted=True)
print(ddf.compute())
```
```
   value
ts      
1    abc
1    abc
2    abc
2    abc
3    abc
3    abc
```
Dask: 2022.5.0
Python: 3.10.4
OS: macOS Monterey 12.1
Install method: pip

_Originally posted by @vdytyniak-exos in https://github.com/dask/dask/issues/8735#issuecomment-1185587580_

### PR DESCRIPTION
**Pull request by jsignell**
- [x] Closes #9339
- [x] Closes #8999
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by pavithraes** [comment]
I've merged main to fix the `test_parquet[fastparquet]` failures :)

**Comment by jsignell** [comment]
I am pretty confident of this change, so I will plan to merge this week unless there are comments.

File: dask/dataframe/tests/test_shuffle.py [hunk 1]
```diff
@@ -1182,6 +1182,17 @@ def test_set_index_overlap_2():
     assert ddf2.npartitions == 8
 
 
+def test_set_index_overlap_3():
```
Comment (ncclementi) [inline dask/dataframe/tests/test_shuffle.py:1185]: I see that we have `test_set_index_overlap` and then `_2` and  `_3`. I thought the only thing changing was the number of partitions but it seems it's not just that. 

Would it be better to have either a comment/docstring to these tests that explains better what each test is accomplishing?

Comment (jsignell) [reply dask/dataframe/tests/test_shuffle.py:1185]: yeah it's a lazy name, I did link out to the issue that has the full context. But I'm happy to change the name

Comment (jsignell) [inline dask/dataframe/tests/test_shuffle.py:1185]:
```suggestion
def test_set_index_overlap_does_not_drop_rows_when_divisions_overlap():
```

**Pull request merged**